content/posts/draft-operator-constraints.markdown

```diff
@@ -0,0 +1,349 @@
+  return func(a, b T) bool { return less(b, a) }
+}
+```
+
+This approach is arguably also more correct than the one above because it
+decouples the type from the comparison used. If I use a `SearchTree` as a set
+datatype, there is no real reason why the elements in the set would be specific
+to the comparison used. It should be “a set of `string`” not “a set of
+`MyCustomlyOrderedString`”. This reflects the fact that with the method
+constraint, we have to convert back-and-forth when putting things into the
+container or taking it out again.
+
+The main *disadvantage* of this approach is that it means you can not have
+useful zero values. Your `SearchTree` type needs the `Less` field to be
+populated to work. So its zero value can not be used to represent an empty set.
+
+You can not even lazily initialize it (which is a common trick to make types
+which need initialization have a useful zero value) because *you don't know
+what it should be*.
+
+## Comparator types
+
+There is a way to pass a function “statically”. That is, instead of passing
+around a `func` value, we can pass it as a type argument. The way to do that is
+to attach it as a method to a `struct{}` type:
+
+```go
+import "golang.org/x/exp/slices"
+
+type IntComparator struct{}
+
+func (IntComparator) Less(a, b int) bool {
+  return a < b
+}
+
+func main() {
+  a := []int{42,23,1337}
+  less := IntComparator{}.Less // has type func(int, int) bool
+  slices.SortFunc(a, less)
+}
+```
+
+Based on this, we can devise a mechanism to allow custom comparisons:
+
+```go
+// Comparator is a helper type used to compare two T values.
+type Comparator[T any] interface {
+  ~struct{}
+  Less(a, b T) bool
+}
+
+func Sort[C Comparator[T], T any](a []T) {
+  var c C
+  less := c.Less // has type func(T, T) bool
+  // …
+}
+
+type SearchTree[C Comparator[T], T any] struct {
+  // …
+}
+```
+
+The `~struct{}` constraints any implementation of `Comparator[T]` to have
+underlying type `struct{}`. It is not strictly necessary, but it serves two
+purposes here:
+
+1. It makes clear that `Comparator[T]` itself is not supposed to carry any
+   state. It only exists to have its method called.
+2. It ensures (as much as possible) that the zero value of `C` is safe to use.
+   In particular, `Comparator[T]` would be a normal interface type. And it
+   would have a `Less` method of the right type, so it would implement itself.
+   But a zero `Comparator[T]` is `nil` and would always panic, if its method is
+   called.
+
+You can again provide helpers. This can also be used to combine this approach
+with the above ones:
+
+```go
+type LessOperator[T constraints.Ordered] struct{}
+
+func (LessOperator[T]) Less(a, b T) bool {
+  return a < b
+}
+
+type LessMethod[T Lesser[T]] struct{}
+
+func (LessMethod[T]) Less(a, b T) bool {
+  return a.Less(b)
+}
+
+type Reversed[C Comparator[T], T any] struct{}
+
+func (Reversed[C, T]) Less(a, b T) bool {
+  var c C
+  return c.Less(b, a)
+}
+```
```

Comment:
Is it possible to define a helper that takes a less function literal?

Reply:
Unfortunately no. Its zero value would not be callable. I will add a sentence to be explicit.